ipc/nano/src/lib.rs

```diff
@@ -199,14 +202,136 @@ impl<S> Worker<S> where S: IpcInterface<S> {
 	}
 }
 
+/// Error in handling JSON RPC request
+pub enum IoHandlerError {
+	BadRequest,
+	HandlerError,
+}
+
+/// Worker to handle JSON RPC requests
+pub struct IoHandlerWorker {
+	handler: Arc<IoHandler>,
+	socket: Socket,
+	_endpoint: Endpoint,
+	poll: Vec<PollFd>,
+	buf: Vec<u8>,
+}
+
+/// IPC server for json-rpc handler (single thread)
+pub struct IoHandlerServer {
+	is_stopping: Arc<AtomicBool>,
+	handler: Arc<IoHandler>,
+	socket_addr: String,
+}
+
+impl IoHandlerServer {
+	/// New IPC server for JSON RPC `handler` and ipc socket address `socket_addr`
+	pub fn new(handler: &Arc<IoHandler>, socket_addr: &str) -> IoHandlerServer {
+		IoHandlerServer {
+			handler: handler.clone(),
+			is_stopping: Arc::new(AtomicBool::new(false)),
+			socket_addr: socket_addr.to_owned(),
+		}
+	}
+
+	/// IPC Server starts (non-blocking, in seprate thread)
+	pub fn start(&self) -> Result<(), SocketError> {
+		let mut worker = try!(IoHandlerWorker::new(&self.handler, &self.socket_addr));
+		self.is_stopping.store(false, Ordering::Relaxed);
+		let worker_is_stopping = self.is_stopping.clone();
+
+		::std::thread::spawn(move || {
+			while !worker_is_stopping.load(Ordering::Relaxed) {
+				worker.poll()
+			}
+		});
+
+		Ok(())
+	}
+
+	/// IPC server will eventually stop
+	pub fn stop(&self) {
```

Comment:
How about implementing `Drop` and stopping the server when reference is dropped?
Or `start()` could return listening handle that when dropped would stop worker.